pkg/token/daemon.go

```diff
@@ -0,0 +1,275 @@
+// Copyright 2023 Yahoo Japan Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package token
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/AthenZ/athenz/clients/go/zts"
+	athenz "github.com/AthenZ/athenz/libs/go/sia/util"
+	"github.com/cenkalti/backoff"
+	"github.com/pkg/errors"
+
+	"github.com/AthenZ/k8s-athenz-sia/v3/pkg/config"
+	extutil "github.com/AthenZ/k8s-athenz-sia/v3/pkg/util"
+	"github.com/AthenZ/k8s-athenz-sia/v3/third_party/log"
+	"github.com/AthenZ/k8s-athenz-sia/v3/third_party/util"
+)
+
+type daemon struct {
+	accessTokenCache TokenCache
+	roleTokenCache   TokenCache
+
+	// keyFile      string
+	// certFile     string
+	// serverCACert string
+	// endpoint     string
+	ztsClient *zts.ZTSClient
+	saService string
+
+	tokenAPIEnable      bool
+	tokenType           Type
+	tokenDir            string
+	tokenRefresh        time.Duration
+	tokenExpiryInSecond int
+	roleAuthHeader      string
+}
+
+func newDaemon(idConfig *config.IdentityConfig, tt Type) (*daemon, error) {
+
+	// initialize token cache with placeholder
+	tokenExpiryInSecond := int(idConfig.TokenExpiry.Seconds())
+	accessTokenCache := NewLockedTokenCache()
+	roleTokenCache := NewLockedTokenCache()
+	targets := strings.Split(idConfig.TargetDomainRoles, ",")
+	if len(targets) != 1 || idConfig.TargetDomainRoles != "" {
+		if tt.Has(ACCESS_TOKEN) {
+			for _, dr := range targets {
+				domain, role, err := athenz.SplitRoleName(dr)
+				if err != nil {
+					return nil, fmt.Errorf("Invalid TargetDomainRoles[%s]: %s", idConfig.TargetDomainRoles, err.Error())
+				}
+				accessTokenCache.Store(CacheKey{Domain: domain, Role: role, MinExpiry: tokenExpiryInSecond}, &AccessToken{})
+			}
+		}
+		if tt.Has(ROLE_TOKEN) {
+			for _, dr := range targets {
+				domain, role, err := athenz.SplitRoleName(dr)
+				if err != nil {
+					return nil, fmt.Errorf("Invalid TargetDomainRoles[%s]: %s", idConfig.TargetDomainRoles, err.Error())
+				}
+				roleTokenCache.Store(CacheKey{Domain: domain, Role: role, MinExpiry: tokenExpiryInSecond}, &RoleToken{})
+			}
+		}
+	}
```

Comment:
```suggestion
	if !tt.HasNone() && len(targets) != 1 || idConfig.TargetDomainRoles != "" {
		for _, dr := range targets {
			domain, role, err := athenz.SplitRoleName(dr)
			if err != nil {
				return nil, fmt.Errorf("Invalid TargetDomainRoles[%s]: %s", idConfig.TargetDomainRoles, err.Error())
			}

			if tt.Has(ACCESS_TOKEN) {
				accessTokenCache.Store(CacheKey{Domain: domain, Role: role, MinExpiry: tokenExpiryInSecond}, &AccessToken{})
			}
			if tt.Has(ROLE_TOKEN) {
				roleTokenCache.Store(CacheKey{Domain: domain, Role: role, MinExpiry: tokenExpiryInSecond}, &RoleToken{})
			}
		}
	}
```
Warning; This may change the original logic. so you need to double check.
HasNone() I just simply created that checks if it has access token token or role token.

Reply:
And if you accept HasAccessToken()
this code will have
`tt.HasAccessToken()` simply.

Reply:
refactor: [634d4eb](https://github.com/WindzCUHK/k8s-athenz-sia/pull/13/commits/634d4eb9b8584c5efe3f99ff4e1d9ae0c7510dbf)

Token type is checked before daemon creation. Either type must exist.
https://github.com/WindzCUHK/k8s-athenz-sia/blob/634d4eb9b8584c5efe3f99ff4e1d9ae0c7510dbf/pkg/token/daemon.go#L208-L214

Reply:
> Token type is checked before daemon creation. Either type must exist.

I see. Thank you. 

> refactor: [634d4eb](https://github.com/WindzCUHK/k8s-athenz-sia/pull/13/commits/634d4eb9b8584c5efe3f99ff4e1d9ae0c7510dbf)

Thank you

Reply:
Resolved